File: handlers/sql_obfuscator.py
import re
import json
import csv
from typing import Dict, List, Set, Tuple
from datetime import datetime


# SQL keywords and types to exclude from columns
SQL_KEYWORDS = {
    # SQL standard keywords
    'select', 'from', 'where', 'and', 'or', 'as', 'on', 'join', 'left', 'right',
    'inner', 'outer', 'full', 'cross', 'group', 'by', 'order', 'having', 'limit',
    'offset', 'union', 'all', 'distinct', 'case', 'when', 'then', 'else', 'end',
    'null', 'true', 'false', 'not', 'in', 'between', 'like', 'is', 'exists',
    'insert', 'update', 'delete', 'create', 'alter', 'drop', 'table', 'index',
    'primary', 'key', 'foreign', 'references', 'constraint', 'default', 'values',
    'set', 'into', 'partition', 'using', 'over', 'window', 'rows', 'range',
    'preceding', 'following', 'current', 'row', 'unbounded', 'asc', 'desc',
    'nulls', 'first', 'last', 'with', 'recursive', 'materialized', 'view',
    'if', 'else', 'cast', 'convert', 'coalesce', 'nullif', 'global', 'local',
    'temporary', 'temp', 'final', 'sample', 'prewhere', 'array', 'tuple',
}

# ClickHouse type patterns (regex)
CLICKHOUSE_TYPE_PATTERN = re.compile(
    r'^(U?Int(8|16|32|64|128|256)|Float(32|64)|Decimal\d*|'
    r'Date(Time(64)?)?|String|FixedString|UUID|Bool|IPv[46]|'
    r'Enum\d*|Array|Tuple|Map|Nullable|LowCardinality|'
    r'SimpleAggregateFunction|AggregateFunction)$',
    re.IGNORECASE
)


def _remove_python_imports(code: str) -> str:
    """
    Remove Python import lines from code to avoid false positives.
    Removes lines starting with 'from X import' or 'import X'.
    """
    lines = code.split('\n')
    filtered_lines = []
    for line in lines:
        stripped = line.strip()
        # Skip Python import lines
        if stripped.startswith('from ') and ' import ' in stripped:
            continue
        if stripped.startswith('import '):
            continue
        filtered_lines.append(line)
    return '\n'.join(filtered_lines)


def _is_sql_keyword_or_type(name: str) -> bool:
    """Check if name is SQL keyword, ClickHouse type, or Python constant."""
    # SQL keywords
    if name.lower() in SQL_KEYWORDS:
        return True
    # ClickHouse types
    if CLICKHOUSE_TYPE_PATTERN.match(name):
        return True
    # Python constants (UPPER_SNAKE_CASE like CLEAR_MANUAL, TABLE_NAME etc.)
    if re.match(r'^[A-Z][A-Z0-9_]*$', name) and len(name) > 2:
        return True
    return False


def extract_entities(code: str) -> Dict[str, Set[str]]:
    """
    Extract all entities from SQL/DAG code.
    Returns dict with sets of found entities by category.
    """
    entities = {
        'schemas': set(),
        'tables': set(),      # includes dictGet targets
        'columns': set(),
        'dag_ids': set(),
        'task_ids': set(),
        'literals': set(),
        'variables': set()    # Python variables (name=value pairs)
    }

    # Remove Python imports before parsing
    code_cleaned = _remove_python_imports(code)

    # Pattern for schema.table from FROM/JOIN (with negative lookahead for 'import')
    pattern_from_join = r'(?:from|join)\s+([a-zA-Z_][a-zA-Z0-9_]*)\.([a-zA-Z_][a-zA-Z0-9_]*)(?!\s+import)'
    matches = re.findall(pattern_from_join, code_cleaned, re.IGNORECASE)
    for schema, table in matches:
        entities['schemas'].add(schema)
        entities['tables'].add(f"{schema}.{table}")

    # Pattern for dictGet('schema.dict')
    pattern_dictget = r"dictGet\s*\(\s*['\"]([a-zA-Z_][a-zA-Z0-9_]*)\.([a-zA-Z_][a-zA-Z0-9_]*)['\"]"
    matches = re.findall(pattern_dictget, code_cleaned, re.IGNORECASE)
    for schema, dict_name in matches:
        entities['schemas'].add(schema)
        entities['tables'].add(f"{schema}.{dict_name}")

    # Pattern for columns from SELECT (only explicit columns)
    pattern_select = r'SELECT\s+(.*?)\s+FROM'
    select_matches = re.findall(pattern_select, code_cleaned, re.IGNORECASE | re.DOTALL)
    for select_clause in select_matches:
        # Remove DISTINCT/ALL/TOP from the beginning
        select_clause = re.sub(r'^\s*(DISTINCT|ALL|TOP\s+\d+)\s+', '', select_clause, flags=re.IGNORECASE)
        # Split by comma, handle aliases
        parts = select_clause.split(',')
        for part in parts:
            part = part.strip()
            # Skip * and complex expressions
            if '*' in part or '(' in part:
                continue
            # Handle alias: col AS alias or col alias
            col_match = re.match(r'^([a-zA-Z_][a-zA-Z0-9_]*(?:\.[a-zA-Z_][a-zA-Z0-9_]*)?)(?:\s+(?:AS\s+)?[a-zA-Z_][a-zA-Z0-9_]*)?$', part, re.IGNORECASE)
            if col_match:
                col = col_match.group(1)
                # If has table prefix (t.column), take only column name
                if '.' in col:
                    col = col.split('.')[-1]
                # Skip SQL keywords, ClickHouse types, and Python constants
                if not _is_sql_keyword_or_type(col):
                    entities['columns'].add(col)

    # Pattern for dag_id
    pattern_dag_id = r"dag_id\s*=\s*['\"]([^'\"]+)['\"]"
    matches = re.findall(pattern_dag_id, code_cleaned, re.IGNORECASE)
    entities['dag_ids'].update(matches)

    # Pattern for task_id
    pattern_task_id = r"task_id\s*=\s*['\"]([^'\"]+)['\"]"
    matches = re.findall(pattern_task_id, code_cleaned, re.IGNORECASE)
    entities['task_ids'].update(matches)

    # Pattern for string literals (length > 3)
    pattern_literals = r"'([^']{4,})'"
    matches = re.findall(pattern_literals, code_cleaned)
    for lit in matches:
        if _is_meaningful_literal(lit):
            entities['literals'].add(lit)

    # Pattern for Python variables
    # First, remove multiline strings content to avoid false positives
    code_no_multiline = re.sub(r'""".*?"""', '""""""', code_cleaned, flags=re.DOTALL)
    code_no_multiline = re.sub(r"'''.*?'''", "''''''", code_no_multiline, flags=re.DOTALL)

    # 1. String assignments: var_name = "value" or var_name = 'value'
    # Capture the quote type to preserve it
    pattern_str_assign = r'([a-zA-Z_][a-zA-Z0-9_]*)\s*=\s*(["\'])([^"\']+)\2'
    matches = re.findall(pattern_str_assign, code_no_multiline)
    for var_name, quote, var_value in matches:
        # Skip short variable names and common temp vars
        if len(var_name) > 2 and var_name.lower() not in ('dag', 'sql', 'tmp', 'var', 'val', 'key', 'row'):
            # Skip dag_id and task_id (already handled separately)
            if var_name.lower() not in ('dag_id', 'task_id'):
                # Format: name|quote|value (use | as separator since = can be in value)
                entities['variables'].add(f"{var_name}|{quote}|{var_value}")

    # 2. UPPER_CASE constants: CONST_NAME = "value"
    pattern_const = r'([A-Z][A-Z0-9_]*)\s*=\s*(["\'])([^"\']+)\2'
    matches = re.findall(pattern_const, code_no_multiline)
    for var_name, quote, var_value in matches:
        if len(var_name) > 2:
            entities['variables'].add(f"{var_name}|{quote}|{var_value}")

    # 3. Multiline string assignments: var = """...""" or var = '''...'''
    # Parse from original code_cleaned (not code_no_multiline)
    # For multiline, only obfuscate the variable NAME, not the value
    pattern_multiline = r'([a-zA-Z_][a-zA-Z0-9_]*)\s*=\s*"""(.*?)"""'
    matches = re.findall(pattern_multiline, code_cleaned, re.DOTALL)
    for var_name, var_value in matches:
        if len(var_name) > 2 and var_value.strip():
            # Store preview for display, mark as multiline with triple quotes
            preview = var_value.strip()[:50].replace('\n', ' ')
            if len(var_value) > 50:
                preview += '...'
            # Use special marker for multiline: name|"""|preview
            entities['variables'].add(f"{var_name}|\"\"\"|{preview}")

    # Also check triple single quotes
    pattern_multiline_sq = r"([a-zA-Z_][a-zA-Z0-9_]*)\s*=\s*'''(.*?)'''"
    matches = re.findall(pattern_multiline_sq, code_cleaned, re.DOTALL)
    for var_name, var_value in matches:
        if len(var_name) > 2 and var_value.strip():
            preview = var_value.strip()[:50].replace('\n', ' ')
            if len(var_value) > 50:
                preview += '...'
            entities['variables'].add(f"{var_name}|'''|{preview}")

    return entities


def _is_meaningful_literal(value: str) -> bool:
    """
    Filter out dates, LIKE patterns, numbers, and SQL keywords.
    """
    # Skip dates (YYYY-MM-DD, DD.MM.YYYY, etc.)
    if re.match(r'^\d{4}-\d{2}-\d{2}', value):
        return False
    if re.match(r'^\d{2}\.\d{2}\.\d{4}', value):
        return False

    # Skip LIKE patterns
    if '%' in value or '_' in value:
        return False

    # Skip pure numbers
    if re.match(r'^\d+$', value):
        return False

    # Skip SQL keywords
    sql_keywords = {'null', 'true', 'false', 'none', 'asc', 'desc'}
    if value.lower() in sql_keywords:
        return False

    # Skip datetime-like strings
    if re.match(r'^\d{4}-\d{2}-\d{2}[T\s]\d{2}:\d{2}', value):
        return False

    return True


def generate_obfuscated_names(entities: Dict[str, Set[str]]) -> List[Dict]:
    """
    Generate obfuscated names for all entities.
    Returns list of mappings with structure:
    {
        'entity_type': str,
        'original_value': str,
        'obfuscated_value': str,
        'enabled': bool
    }
    """
    mappings = []
    counters = {
        'schema': 1,
        'table': 1,
        'column': 1,
        'dag': 1,
        'task': 1,
        'literal': 1,
        'variable': 1
    }

    # Map schemas first (needed for table obfuscation)
    schema_map = {}
    for schema in sorted(entities.get('schemas', [])):
        obf_name = f"sch_{counters['schema']:03d}"
        schema_map[schema] = obf_name
        mappings.append({
            'entity_type': 'schema',
            'original_value': schema,
            'obfuscated_value': obf_name,
            'enabled': True
        })
        counters['schema'] += 1

    # Tables (schema.table -> sch_XXX.obj_YYY)
    for table in sorted(entities.get('tables', [])):
        if '.' in table:
            schema, tbl_name = table.split('.', 1)
            obf_schema = schema_map.get(schema, schema)
            obf_table = f"obj_{counters['table']:03d}"
            obf_full = f"{obf_schema}.{obf_table}"
        else:
            obf_full = f"obj_{counters['table']:03d}"
        mappings.append({
            'entity_type': 'table',
            'original_value': table,
            'obfuscated_value': obf_full,
            'enabled': True
        })
        counters['table'] += 1

    # Columns
    for col in sorted(entities.get('columns', [])):
        obf_name = f"col_{counters['column']:03d}"
        mappings.append({
            'entity_type': 'column',
            'original_value': col,
            'obfuscated_value': obf_name,
            'enabled': True
        })
        counters['column'] += 1

    # DAG IDs
    for dag_id in sorted(entities.get('dag_ids', [])):
        obf_name = f"dag_{counters['dag']:03d}"
        mappings.append({
            'entity_type': 'dag',
            'original_value': dag_id,
            'obfuscated_value': obf_name,
            'enabled': True
        })
        counters['dag'] += 1

    # Task IDs
    for task_id in sorted(entities.get('task_ids', [])):
        obf_name = f"task_{counters['task']:03d}"
        mappings.append({
            'entity_type': 'task',
            'original_value': task_id,
            'obfuscated_value': obf_name,
            'enabled': True
        })
        counters['task'] += 1

    # Literals
    for lit in sorted(entities.get('literals', [])):
        obf_name = f"str_{counters['literal']:03d}"
        mappings.append({
            'entity_type': 'literal',
            'original_value': lit,
            'obfuscated_value': obf_name,
            'enabled': False  # Literals disabled by default
        })
        counters['literal'] += 1

    # Variables (format: "var_name|quote|value")
    for var_entry in sorted(entities.get('variables', [])):
        if '|' in var_entry:
            parts = var_entry.split('|', 2)
            if len(parts) == 3:
                var_name, quote, var_value = parts
                obf_var_name = f"var_{counters['variable']:03d}"

                # Check if multiline (triple quotes) - only obfuscate name
                is_multiline = quote in ('"""', "'''")

                if is_multiline:
                    # For multiline: only change variable name, keep value
                    # Display format: name = """preview..."""
                    display_orig = f"{var_name} = {quote}{var_value}{quote}"
                    display_obf = f"{obf_var_name} = {quote}(unchanged){quote}"
                    mappings.append({
                        'entity_type': 'variable',
                        'original_value': var_entry,
                        'original_display': display_orig,
                        'obfuscated_value': f"{obf_var_name}|{quote}|",  # Empty value = keep original
                        'obfuscated_display': display_obf,
                        'enabled': False
                    })
                else:
                    # For regular strings: obfuscate both name and value
                    obf_var_value = f"val_{counters['variable']:03d}"
                    display_orig = f"{var_name} = {quote}{var_value}{quote}"
                    display_obf = f"{obf_var_name} = {quote}{obf_var_value}{quote}"
                    mappings.append({
                        'entity_type': 'variable',
                        'original_value': var_entry,
                        'original_display': display_orig,
                        'obfuscated_value': f"{obf_var_name}|{quote}|{obf_var_value}",
                        'obfuscated_display': display_obf,
                        'enabled': False
                    })
                counters['variable'] += 1

    return mappings


def apply_replacements(code: str, mappings: List[Dict]) -> str:
    """
    Apply enabled replacements to code.
    Order: tables first (longer matches), then schemas, columns, etc.
    """
    result = code

    # Sort mappings: longer original values first to avoid partial replacements
    enabled_mappings = [m for m in mappings if m.get('enabled', True)]
    enabled_mappings.sort(key=lambda m: len(m['original_value']), reverse=True)

    # Group by type for ordered replacement
    by_type = {}
    for m in enabled_mappings:
        t = m['entity_type']
        if t not in by_type:
            by_type[t] = []
        by_type[t].append(m)

    # Replace tables first (full schema.table)
    for m in by_type.get('table', []):
        # Use word boundaries for table names
        pattern = re.escape(m['original_value'])
        result = re.sub(
            r'(?<![a-zA-Z0-9_])' + pattern + r'(?![a-zA-Z0-9_])',
            m['obfuscated_value'],
            result,
            flags=re.IGNORECASE
        )

    # Replace DAG IDs (in quotes)
    for m in by_type.get('dag', []):
        pattern = r"(dag_id\s*=\s*['\"])" + re.escape(m['original_value']) + r"(['\"])"
        result = re.sub(pattern, r'\1' + m['obfuscated_value'] + r'\2', result, flags=re.IGNORECASE)

    # Replace Task IDs (in quotes)
    for m in by_type.get('task', []):
        pattern = r"(task_id\s*=\s*['\"])" + re.escape(m['original_value']) + r"(['\"])"
        result = re.sub(pattern, r'\1' + m['obfuscated_value'] + r'\2', result, flags=re.IGNORECASE)

    # Replace columns (with word boundaries)
    for m in by_type.get('column', []):
        pattern = r'(?<![a-zA-Z0-9_])' + re.escape(m['original_value']) + r'(?![a-zA-Z0-9_])'
        result = re.sub(pattern, m['obfuscated_value'], result)

    # Replace literals (in quotes)
    for m in by_type.get('literal', []):
        pattern = r"'" + re.escape(m['original_value']) + r"'"
        result = re.sub(pattern, "'" + m['obfuscated_value'] + "'", result)

    # Replace variables (format: name|quote|value)
    for m in by_type.get('variable', []):
        orig = m['original_value']
        obf = m['obfuscated_value']
        if '|' in orig and '|' in obf:
            orig_parts = orig.split('|', 2)
            obf_parts = obf.split('|', 2)
            if len(orig_parts) == 3 and len(obf_parts) == 3:
                orig_name, orig_quote, orig_value = orig_parts
                obf_name, obf_quote, obf_value = obf_parts

                # Always replace variable name (with word boundaries)
                pattern = r'(?<![a-zA-Z0-9_])' + re.escape(orig_name) + r'(?![a-zA-Z0-9_])'
                result = re.sub(pattern, obf_name, result)

                # Replace value only if obf_value is not empty (not multiline)
                if obf_value:
                    # For single/double quoted strings - use re.sub for proper escaping
                    if orig_quote in ('"', "'"):
                        pattern = re.escape(orig_quote) + re.escape(orig_value) + re.escape(orig_quote)
                        replacement = orig_quote + obf_value + orig_quote
                        result = re.sub(pattern, replacement, result)

    return result


def generate_session_name() -> str:
    """Generate auto session name."""
    return datetime.now().strftime("session_%Y-%m-%d_%H-%M")


def export_to_json(mappings: List[Dict], filepath: str):
    """Export mappings to JSON file."""
    # Remove 'enabled' field for export
    export_data = [
        {
            'entity_type': m['entity_type'],
            'original_value': m['original_value'],
            'obfuscated_value': m['obfuscated_value']
        }
        for m in mappings if m.get('enabled', True)
    ]
    with open(filepath, 'w', encoding='utf-8') as f:
        json.dump(export_data, f, ensure_ascii=False, indent=2)


def export_to_csv(mappings: List[Dict], filepath: str):
    """Export mappings to CSV file."""
    with open(filepath, 'w', encoding='utf-8', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(['entity_type', 'original_value', 'obfuscated_value'])
        for m in mappings:
            if m.get('enabled', True):
                writer.writerow([m['entity_type'], m['original_value'], m['obfuscated_value']])


def load_from_file(filepath: str) -> List[Dict]:
    """Load mappings from JSON or CSV file."""
    mappings = []

    if filepath.endswith('.json'):
        with open(filepath, 'r', encoding='utf-8') as f:
            data = json.load(f)
            for item in data:
                mappings.append({
                    'entity_type': item['entity_type'],
                    'original_value': item['original_value'],
                    'obfuscated_value': item['obfuscated_value'],
                    'enabled': True
                })
    elif filepath.endswith('.csv'):
        with open(filepath, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            for row in reader:
                mappings.append({
                    'entity_type': row['entity_type'],
                    'original_value': row['original_value'],
                    'obfuscated_value': row['obfuscated_value'],
                    'enabled': True
                })

    return mappings

File: desktop-rust/src-tauri/src/handlers/sql_obfuscator.rs
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::{BTreeSet, HashMap};

/// A single mapping entry from original to obfuscated name.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ObfuscationEntry {
    pub entity_type: String,
    pub original_value: String,
    pub obfuscated_value: String,
    pub enabled: bool,
}

/// SQL keywords to exclude from column detection.
const SQL_KEYWORDS: &[&str] = &[
    "select", "from", "where", "and", "or", "as", "on", "join", "left", "right",
    "inner", "outer", "full", "cross", "group", "by", "order", "having", "limit",
    "offset", "union", "all", "distinct", "case", "when", "then", "else", "end",
    "null", "true", "false", "not", "in", "between", "like", "is", "exists",
    "insert", "update", "delete", "create", "alter", "drop", "table", "index",
    "primary", "key", "foreign", "references", "constraint", "default", "values",
    "set", "into", "partition", "using", "over", "window", "rows", "range",
    "preceding", "following", "current", "row", "unbounded", "asc", "desc",
    "nulls", "first", "last", "with", "recursive", "materialized", "view",
    "if", "cast", "convert", "coalesce", "nullif", "global", "local",
    "temporary", "temp", "final", "sample", "prewhere", "array", "tuple",
];

fn is_sql_keyword(word: &str) -> bool {
    SQL_KEYWORDS.contains(&word.to_lowercase().as_str())
}

/// Remove Python import lines from code.
fn remove_python_imports(code: &str) -> String {
    code.lines()
        .filter(|line| {
            let trimmed = line.trim();
            !(trimmed.starts_with("from ") && trimmed.contains(" import "))
                && !trimmed.starts_with("import ")
        })
        .collect::<Vec<_>>()
        .join("\n")
}

/// Extract all entities (schemas, tables, columns, dag_ids, task_ids, literals) from SQL/DAG code.
pub fn extract_entities(code: &str) -> HashMap<String, BTreeSet<String>> {
    let mut entities: HashMap<String, BTreeSet<String>> = HashMap::new();
    entities.insert("schemas".into(), BTreeSet::new());
    entities.insert("tables".into(), BTreeSet::new());
    entities.insert("columns".into(), BTreeSet::new());
    entities.insert("dag_ids".into(), BTreeSet::new());
    entities.insert("task_ids".into(), BTreeSet::new());
    entities.insert("literals".into(), BTreeSet::new());

    let cleaned = remove_python_imports(code);

    // Pattern for schema.table from FROM/JOIN
    // Note: Python imports are already filtered by remove_python_imports, so no lookahead needed
    let re_from_join = Regex::new(
        r"(?i)\b(?:from|join)\s+([a-zA-Z_][a-zA-Z0-9_]*)\.([a-zA-Z_][a-zA-Z0-9_]*)"
    ).unwrap();
    for cap in re_from_join.captures_iter(&cleaned) {
        let schema = cap[1].to_string();
        let table = cap[2].to_string();
        entities.get_mut("schemas").unwrap().insert(schema.clone());
        entities.get_mut("tables").unwrap().insert(format!("{schema}.{table}"));
    }

    // Pattern for dictGet('schema.dict')
    let re_dictget = Regex::new(
        r#"(?i)dictGet\s*\(\s*['"]([a-zA-Z_][a-zA-Z0-9_]*)\.([a-zA-Z_][a-zA-Z0-9_]*)['"]"#
    ).unwrap();
    for cap in re_dictget.captures_iter(&cleaned) {
        let schema = cap[1].to_string();
        let dict_name = cap[2].to_string();
        entities.get_mut("schemas").unwrap().insert(schema.clone());
        entities.get_mut("tables").unwrap().insert(format!("{schema}.{dict_name}"));
    }

    // Pattern for columns from SELECT ... FROM
    let re_select = Regex::new(r"(?is)SELECT\s+(.*?)\s+FROM").unwrap();
    let re_col = Regex::new(
        r"(?i)^([a-zA-Z_][a-zA-Z0-9_]*(?:\.[a-zA-Z_][a-zA-Z0-9_]*)?)(?:\s+(?:AS\s+)?[a-zA-Z_][a-zA-Z0-9_]*)?$"
    ).unwrap();
    for cap in re_select.captures_iter(&cleaned) {
        let select_clause = &cap[1];
        // Remove DISTINCT/ALL/TOP from beginning
        let re_prefix = Regex::new(r"(?i)^\s*(DISTINCT|ALL|TOP\s+\d+)\s+").unwrap();
        let select_clause = re_prefix.replace(select_clause, "").to_string();
        for part in select_clause.split(',') {
            let part = part.trim();
            if part.contains('*') || part.contains('(') {
                continue;
            }
            if let Some(col_cap) = re_col.captures(part) {
                let mut col = col_cap[1].to_string();
                if col.contains('.') {
                    col = col.split('.').last().unwrap_or(&col).to_string();
                }
                if !is_sql_keyword(&col) && col.len() > 1 {
                    entities.get_mut("columns").unwrap().insert(col);
                }
            }
        }
    }

    // Pattern for dag_id
    let re_dag = Regex::new(r#"(?i)dag_id\s*=\s*['"]([^'"]+)['"]"#).unwrap();
    for cap in re_dag.captures_iter(&cleaned) {
        entities.get_mut("dag_ids").unwrap().insert(cap[1].to_string());
    }

    // Pattern for task_id
    let re_task = Regex::new(r#"(?i)task_id\s*=\s*['"]([^'"]+)['"]"#).unwrap();
    for cap in re_task.captures_iter(&cleaned) {
        entities.get_mut("task_ids").unwrap().insert(cap[1].to_string());
    }

    // Pattern for string literals (length > 3)
    let re_lit = Regex::new(r"'([^']{4,})'").unwrap();
    for cap in re_lit.captures_iter(&cleaned) {
        let lit = &cap[1];
        if is_meaningful_literal(lit) {
            entities.get_mut("literals").unwrap().insert(lit.to_string());
        }
    }

    entities
}

fn is_meaningful_literal(value: &str) -> bool {
    // Skip dates
    let re_date1 = Regex::new(r"^\d{4}-\d{2}-\d{2}").unwrap();
    let re_date2 = Regex::new(r"^\d{2}\.\d{2}\.\d{4}").unwrap();
    if re_date1.is_match(value) || re_date2.is_match(value) {
        return false;
    }
    // Skip LIKE patterns
    if value.contains('%') || value.contains('_') {
        return false;
    }
    // Skip pure numbers
    if value.chars().all(|c| c.is_ascii_digit()) {
        return false;
    }
    // Skip SQL constants
    let skip = ["null", "true", "false", "none", "asc", "desc"];
    if skip.contains(&value.to_lowercase().as_str()) {
        return false;
    }
    true
}

/// Generate obfuscated names for all extracted entities.
pub fn generate_obfuscated_names(entities: &HashMap<String, BTreeSet<String>>) -> Vec<ObfuscationEntry> {
    let mut mappings = Vec::new();
    let mut schema_counter = 1u32;
    let mut table_counter = 1u32;
    let mut col_counter = 1u32;
    let mut dag_counter = 1u32;
    let mut task_counter = 1u32;
    let mut literal_counter = 1u32;

    let empty = BTreeSet::new();

    // Schemas
    let mut schema_map: HashMap<String, String> = HashMap::new();
    for schema in entities.get("schemas").unwrap_or(&empty) {
        let obf = format!("sch_{schema_counter:03}");
        schema_map.insert(schema.clone(), obf.clone());
        mappings.push(ObfuscationEntry {
            entity_type: "schema".into(),
            original_value: schema.clone(),
            obfuscated_value: obf,
            enabled: true,
        });
        schema_counter += 1;
    }

    // Tables (schema.table -> sch_XXX.obj_YYY)
    for table in entities.get("tables").unwrap_or(&empty) {
        let obf = if table.contains('.') {
            let parts: Vec<&str> = table.splitn(2, '.').collect();
            let obf_schema = schema_map.get(parts[0]).cloned().unwrap_or_else(|| parts[0].to_string());
            format!("{obf_schema}.obj_{table_counter:03}")
        } else {
            format!("obj_{table_counter:03}")
        };
        mappings.push(ObfuscationEntry {
            entity_type: "table".into(),
            original_value: table.clone(),
            obfuscated_value: obf,
            enabled: true,
        });
        table_counter += 1;
    }

    // Columns
    for col in entities.get("columns").unwrap_or(&empty) {
        mappings.push(ObfuscationEntry {
            entity_type: "column".into(),
            original_value: col.clone(),
            obfuscated_value: format!("col_{col_counter:03}"),
            enabled: true,
        });
        col_counter += 1;
    }

    // DAG IDs
    for dag_id in entities.get("dag_ids").unwrap_or(&empty) {
        mappings.push(ObfuscationEntry {
            entity_type: "dag".into(),
            original_value: dag_id.clone(),
            obfuscated_value: format!("dag_{dag_counter:03}"),
            enabled: true,
        });
        dag_counter += 1;
    }

    // Task IDs
    for task_id in entities.get("task_ids").unwrap_or(&empty) {
        mappings.push(ObfuscationEntry {
            entity_type: "task".into(),
            original_value: task_id.clone(),
            obfuscated_value: format!("task_{task_counter:03}"),
            enabled: true,
        });
        task_counter += 1;
    }

    // Literals (disabled by default)
    for lit in entities.get("literals").unwrap_or(&empty) {
        mappings.push(ObfuscationEntry {
            entity_type: "literal".into(),
            original_value: lit.clone(),
            obfuscated_value: format!("str_{literal_counter:03}"),
            enabled: false,
        });
        literal_counter += 1;
    }

    mappings
}

/// Apply enabled replacements to code. Longer matches are replaced first.
pub fn apply_replacements(code: &str, mappings: &[ObfuscationEntry]) -> String {
    let mut result = code.to_string();

    // Sort: longer original values first to avoid partial replacements
    let mut enabled: Vec<_> = mappings.iter().filter(|m| m.enabled).collect();
    enabled.sort_by(|a, b| b.original_value.len().cmp(&a.original_value.len()));

    // Group by type
    let mut by_type: HashMap<&str, Vec<&&ObfuscationEntry>> = HashMap::new();
    for m in &enabled {
        by_type.entry(m.entity_type.as_str()).or_default().push(m);
    }

    // Replace tables first (schema.table — longest)
    for m in by_type.get("table").unwrap_or(&vec![]) {
        // Use \b for word boundary (works for schema.table since . is not a word char)
        let pattern = format!(
            r"(?i)\b{}\b",
            regex::escape(&m.original_value)
        );
        if let Ok(re) = Regex::new(&pattern) {
            result = re.replace_all(&result, m.obfuscated_value.as_str()).to_string();
        }
    }

    // Replace DAG IDs (in quotes)
    for m in by_type.get("dag").unwrap_or(&vec![]) {
        let pattern = format!(
            r#"(?i)(dag_id\s*=\s*['"]){}(['"])"#,
            regex::escape(&m.original_value)
        );
        if let Ok(re) = Regex::new(&pattern) {
            let replacement = format!("${{1}}{}${{2}}", m.obfuscated_value);
            result = re.replace_all(&result, replacement.as_str()).to_string();
        }
    }

    // Replace Task IDs (in quotes)
    for m in by_type.get("task").unwrap_or(&vec![]) {
        let pattern = format!(
            r#"(?i)(task_id\s*=\s*['"]){}(['"])"#,
            regex::escape(&m.original_value)
        );
        if let Ok(re) = Regex::new(&pattern) {
            let replacement = format!("${{1}}{}${{2}}", m.obfuscated_value);
            result = re.replace_all(&result, replacement.as_str()).to_string();
        }
    }

    // Replace columns (with word boundaries)
    for m in by_type.get("column").unwrap_or(&vec![]) {
        let pattern = format!(
            r"\b{}\b",
            regex::escape(&m.original_value)
        );
        if let Ok(re) = Regex::new(&pattern) {
            result = re.replace_all(&result, m.obfuscated_value.as_str()).to_string();
        }
    }

    // Replace literals (in quotes)
    for m in by_type.get("literal").unwrap_or(&vec![]) {
        let pattern = format!("'{}'", regex::escape(&m.original_value));
        if let Ok(re) = Regex::new(&pattern) {
            let replacement = format!("'{}'", m.obfuscated_value);
            result = re.replace_all(&result, replacement.as_str()).to_string();
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_tables() {
        let entities = extract_entities("SELECT a FROM mydb.orders JOIN mydb.users ON x=y");
        let tables = entities.get("tables").unwrap();
        assert!(tables.contains("mydb.orders"));
        assert!(tables.contains("mydb.users"));
    }

    #[test]
    fn test_extract_dictget() {
        let entities = extract_entities("SELECT dictGet('mydb.my_dict', 'col', id) FROM mydb.t1");
        let tables = entities.get("tables").unwrap();
        assert!(tables.contains("mydb.my_dict"));
        assert!(tables.contains("mydb.t1"));
    }

    #[test]
    fn test_generate_obfuscated_names() {
        let entities = extract_entities("SELECT col1, col2 FROM schema1.table1 JOIN schema1.table2 ON x=y");
        let mappings = generate_obfuscated_names(&entities);
        assert!(!mappings.is_empty());
        // Should have schema, table, and column entries
        let types: BTreeSet<_> = mappings.iter().map(|m| m.entity_type.as_str()).collect();
        assert!(types.contains("schema"));
        assert!(types.contains("table"));
        assert!(types.contains("column"));
    }

    #[test]
    fn test_apply_replacements() {
        let code = "SELECT col1 FROM mydb.orders WHERE col1 > 0";
        let mappings = vec![
            ObfuscationEntry {
                entity_type: "table".into(),
                original_value: "mydb.orders".into(),
                obfuscated_value: "sch_001.obj_001".into(),
                enabled: true,
            },
            ObfuscationEntry {
                entity_type: "column".into(),
                original_value: "col1".into(),
                obfuscated_value: "col_001".into(),
                enabled: true,
            },
        ];
        let result = apply_replacements(code, &mappings);
        assert!(result.contains("sch_001.obj_001"));
        assert!(result.contains("col_001"));
        assert!(!result.contains("mydb.orders"));
    }

    #[test]
    fn test_python_imports_removed() {
        let code = "from module import func\nSELECT * FROM db.table1";
        let entities = extract_entities(code);
        let tables = entities.get("tables").unwrap();
        assert!(tables.contains("db.table1"));
        assert!(!tables.iter().any(|t| t.contains("module")));
    }
}
